File: bindings/cpp/examples/device/camera_record.cpp
// camera_record.cpp - device capability: camera + mic capture -> H.264 + AAC
// encode -> single two-track MP4.
//
// Status: real. Camera capture (CPU NV12 frames), auto H.264 encode, and the
// container remux all run against the shipped C ABI underneath this example;
// the audio encoder is ABI v2 (adr/0003-auto-audio-encode-c-abi.md), so the
// microphone PCM is now ENCODED to AAC and muxed, not drained (the drain-only
// gap the old example carried is gone).
//
// Flow:
//   1. Open camera index 0 at 30 fps - graceful (skip) if absent.
//   2. Open the microphone at 48 kHz - graceful (video-only) if absent.
//   3. Read the negotiated capture geometry + mic format and open the auto
//      encoders at the REAL negotiated values.
//   4. Record ~3 s: poll video frames into the video encode session, push mic
//      PCM into the audio encode session.
//   5. finish() the video session -> fMP4; flush the audio session -> AAC
//      packets. REMUX: demux the video fMP4, mux video + AAC audio into one
//      two-track fMP4 (the audio track registered with the encoder's
//      AudioSpecificConfig). Without audio, write the video-only fMP4.
//   6. close() the captures (joins their worker threads); the result is
//      written to camera_out.mp4 by the caller.

#include <mediaway/mediaway.hpp>

#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <optional>
#include <string>
#include <thread>
#include <vector>

namespace {

// NO_BACKEND / UNSUPPORTED / NO_DEVICE are expected outcomes in the ABI, not
// hard failures: the caller degrades gracefully (video-only, or skip capture).
bool isExpectedUnavailable(const mediaway::Error& error) {
  return error.status() == mediaway::Status::NoBackend ||
         error.status() == mediaway::Status::Unsupported ||
         error.status() == mediaway::Status::NoDevice;
}

bool writeFile(const std::string& path, const mediaway::Bytes& bytes) {
  std::ofstream out(path, std::ios::binary);
  if (!out) return false;
  out.write(reinterpret_cast<const char*>(bytes.data()),
            static_cast<std::streamsize>(bytes.size()));
  return static_cast<bool>(out);
}

}  // namespace

int main() {
  try {
    // ---- Camera: index 0, 30 fps; 0x0 geometry = pick the camera default ----
    std::optional<mediaway::device::VideoCapture> camera;
    try {
      camera.emplace(mediaway::device::VideoCapture::open({0, {1, 30}}));
    } catch (const mediaway::Error& error) {
      if (!isExpectedUnavailable(error)) throw;
      std::cerr << "camera unavailable: " << error.what()
                << " - nothing to record\n";
      return EXIT_SUCCESS;  // graceful: no camera is a normal situation
    }

    // ---- Microphone: 48 kHz requested; negotiated format is authoritative ---
    std::optional<mediaway::device::AudioCapture> mic;
    try {
      mic.emplace(mediaway::device::AudioCapture::open({0, 48000}));
    } catch (const mediaway::Error& error) {
      if (!isExpectedUnavailable(error)) throw;
      std::cerr << "microphone unavailable: " << error.what()
                << " - recording video only\n";
    }

    // ---- Negotiated geometry -> encoder at the REAL resolution --------------
    const mediaway::device::CaptureInfo& info = camera->info();
    std::cout << "camera negotiated " << info.width << 'x' << info.height
              << " @ " << info.frameRate.num << '/' << info.frameRate.den
              << '\n';

    mediaway::encoder::AutoVideoEncoder encoder =
        mediaway::encoder::AutoVideoEncoder::open({
            mediaway::Codec::H264,
            info.width, info.height,
            info.frameRate,
            info.format,  // the camera delivers NV12 CPU frames
        });
    mediaway::encoder::EncodeSession session = std::move(encoder).begin();

    // ---- Audio encoder at the mic's negotiated format -----------------------
    std::optional<mediaway::encoder::AudioEncoder> audioEncoder;
    if (mic) {
      std::uint32_t sampleRate = 0;
      std::uint16_t channels = 0;
      mic->format(sampleRate, channels);
      std::cout << "mic negotiated " << sampleRate << " Hz, " << channels
                << " channel(s)\n";
      try {
        audioEncoder.emplace(mediaway::encoder::AudioEncoder::open(
            sampleRate, channels, {1, sampleRate}));
      } catch (const mediaway::Error& error) {
        if (!isExpectedUnavailable(error)) throw;
        std::cerr << "audio encoder unavailable: " << error.what()
                  << " - recording video only\n";
      }
    }

    // ---- Record ~3 s ----------------------------------------------------------
    const auto deadline =
        std::chrono::steady_clock::now() + std::chrono::seconds(3);
    std::size_t videoFrames = 0;
    std::size_t audioFrames = 0;
    while (std::chrono::steady_clock::now() < deadline) {
      if (std::optional<mediaway::VideoFrame> frame = camera->pollFrame()) {
        session.writeFrame(*frame);
        ++videoFrames;
      }
      if (mic && audioEncoder) {
        while (std::optional<mediaway::device::AudioFrame> audio =
                   mic->pollFrame()) {
          audioEncoder->pushPcm(*audio);
          ++audioFrames;
        }
      }
      std::this_thread::sleep_for(std::chrono::milliseconds(2));
    }

    // Explicit close joins the capture worker threads (can block up to one
    // frame interval); the destructors would do the same, but finishing the
    // encode with captures stopped is tidier.
    camera->close();
    if (mic) mic->close();

    // ---- Finish encodes -------------------------------------------------------
    mediaway::Bytes mp4 = std::move(session).finish();

    std::vector<mediaway::Packet> audioPackets;
    std::optional<mediaway::AudioStreamInfo> audioInfo;
    if (audioEncoder) {
      audioEncoder->flush();
      while (std::optional<mediaway::Packet> packet = audioEncoder->pollPacket()) {
        audioPackets.push_back(std::move(*packet));
      }
      audioInfo = audioEncoder->streamInfo();  // ASC materialized after the first push
    }
    const bool haveAudio = audioPackets.size() > 0 && audioInfo.has_value() &&
                           !audioInfo->codecConfig.empty();

    // ---- Remux video + AAC into one two-track MP4 -----------------------------
    mediaway::Bytes out;
    if (haveAudio) {
      mediaway::container::Demuxer demuxer;
      demuxer.pushBytes(mp4);
      std::vector<mediaway::StreamInfo> streams = demuxer.streams();
      if (streams.size() != 1 ||
          !std::holds_alternative<mediaway::VideoStreamInfo>(streams[0])) {
        std::cerr << "expected exactly one video stream from the encode "
                     "session's fMP4\n";
        return EXIT_FAILURE;
      }
      const mediaway::VideoStreamInfo& vinfo =
          std::get<mediaway::VideoStreamInfo>(streams[0]);

      mediaway::container::Muxer muxer;
      const mediaway::TrackId videoTrack = muxer.addVideoTrack(vinfo);
      const mediaway::TrackId audioTrack = muxer.addAudioTrack(*audioInfo);
      mediaway::container::LiveMuxer live = std::move(muxer).begin();

      while (std::optional<mediaway::Packet> packet = demuxer.pollPacket()) {
        packet->trackId = videoTrack;  // remap to the new muxer's registration
        live.pushPacket(*packet);
      }
      for (mediaway::Packet& packet : audioPackets) {
        packet.trackId = audioTrack;
        live.pushPacket(packet);
      }
      live.flush();
      out = live.pollBytes();
    } else {
      out = std::move(mp4);
    }

    std::cout << "recorded " << videoFrames << " video frames";
    if (haveAudio) {
      std::cout << " from " << audioFrames << " PCM frame(s) -> "
                << audioPackets.size() << " AAC packets ("
                << audioInfo->sampleRate << " Hz, " << audioInfo->channels
                << " ch)";
    } else {
      std::cout << " (audio unavailable)";
    }
    std::cout << " -> " << out.size() << " bytes\n";
    if (!writeFile("camera_out.mp4", out)) {
      std::cerr << "failed to write camera_out.mp4\n";
      return EXIT_FAILURE;
    }
    std::cout << "wrote camera_out.mp4\n";
    return EXIT_SUCCESS;
  } catch (const mediaway::Error& error) {
    std::cerr << "mediaway error: " << error.what() << " (status "
              << static_cast<int>(error.status()) << ")\n";
    return EXIT_FAILURE;
  }
}

File: bindings/python/examples/device/camera_record.py
"""Record ~3 seconds from the default camera and microphone, encode to H.264 +
AAC, and remux both into one two-track MP4 (out.mp4).

✅ REAL — camera capture, auto H.264 encode, audio encode (ABI v2, adr/0003)
and the container remux are all implemented in the native C ABI
(`mediaway_device_ffi` + `mediaway_pipeline_ffi` + `mediaway_container_ffi`);
this example runs against them. The old "drained, not muxed" gap is gone.

Flow: camera frames go through the auto video encoder's internal MP4 session
(video-only by design); mic PCM goes through AudioEncoder (single-step open —
the session IS the encoder). Then we REMUX: demux the video session's fMP4 and
mux video + AAC audio into one two-track fMP4, registering the audio track
with the encoder's AudioSpecificConfig (`stream_info()`, available after the
first pushed frame). Without a mic/audio backend, the result is video-only.
Missing hardware degrades gracefully: no status-code checks, just exceptions.
"""

import dataclasses
from contextlib import nullcontext
from pathlib import Path
import time

from mediaway import (
    AudioCapture,
    AudioEncoder,
    AutoVideoEncoder,
    Demuxer,
    DeviceUnavailableError,
    EncodeSession,
    EncoderUnavailableError,
    Muxer,
    Rational,
    VideoCapture,
    VideoStreamInfo,
)

RECORD_SECONDS = 3.0
MIC_DRAIN_SECONDS = 0.25  # encode a short window of buffered PCM after recording


def fmt_rational(r: Rational) -> str:
    """Render a Rational as a compact "num/den" string."""
    return f"{r.num}/{r.den}"


def main() -> None:
    try:
        camera = VideoCapture.open("camera", index=0, frame_rate=Rational(1, 30))
    except DeviceUnavailableError as err:
        print(f"camera unavailable: {err}; nothing to record")
        return

    with camera:
        # The capture backend negotiates the real geometry, which may differ
        # from what we asked for — encode at what we actually got.
        width, height = camera.size()
        fps = camera.frame_rate()
        print(f"negotiated geometry: {width}x{height} @ {fmt_rational(fps)}")

        try:
            encoder = AutoVideoEncoder.pick(
                width=width, height=height, frame_rate=fps
            )
        except EncoderUnavailableError as err:
            print(f"no encoder available for {width}x{height}: {err}")
            return
        print(f"picked encoder: {encoder.name} ({encoder.codec.name})")

        # Mic is optional: without it we record video only.
        mic = None
        audio_encoder = None
        try:
            mic = AudioCapture.open("mic", 48000)
            print(f"mic open: {mic.sample_rate()} Hz, {mic.channels()} ch")
        except DeviceUnavailableError as err:
            print(f"mic unavailable ({err}); recording video only")

        if mic is not None:
            try:
                # Encode at the mic's NEGOTIATED format (a mono mic is not
                # the AAC sugar's default stereo).
                audio_encoder = AudioEncoder.open(
                    sample_rate=mic.sample_rate(), channels=mic.channels()
                )
            except EncoderUnavailableError as err:
                print(f"no audio encoder ({err}); recording video only")
                audio_encoder = None

        with (
            mic or nullcontext(),
            EncodeSession(encoder) as session,
            audio_encoder or nullcontext(),
        ):
            deadline = time.monotonic() + RECORD_SECONDS
            frames_recorded = 0
            pcm_bytes = 0
            while time.monotonic() < deadline:
                frame = camera.poll_frame(timeout=0.05)
                if frame is not None:
                    session.push_frame(frame)
                    frames_recorded += 1

                if audio_encoder is not None:
                    pcm = mic.poll_pcm(timeout=0.0)
                    if pcm is not None:
                        audio_encoder.push_pcm(pcm)  # pts auto-advances
                        pcm_bytes += len(pcm)

            # Encode whatever the mic still had buffered during the loop.
            if audio_encoder is not None:
                drain_until = time.monotonic() + MIC_DRAIN_SECONDS
                while time.monotonic() < drain_until:
                    pcm = mic.poll_pcm(timeout=0.05)
                    if pcm is not None:
                        audio_encoder.push_pcm(pcm)
                        pcm_bytes += len(pcm)

            # Audio encode finish must happen inside the `with` block — the
            # context manager closes the encoder on exit (close is always safe,
            # but a closed handle rejects further calls).
            audio_packets = []
            audio_info = None
            if audio_encoder is not None:
                audio_encoder.flush()
                while True:
                    packet = audio_encoder.poll_packet()
                    if packet is None:
                        break
                    audio_packets.append(packet)
                audio_info = audio_encoder.stream_info()  # ASC materialized after the first push

            mp4 = session.finish()  # terminal: consumes the session

        # ---- Remux video + AAC into one two-track MP4 ------------------------
        have_audio = bool(audio_packets) and bool(audio_info and audio_info.extra_data)

        if have_audio:
            demuxer = Demuxer()
            demuxer.push_bytes(mp4)
            streams = demuxer.streams()
            if len(streams) != 1 or not isinstance(streams[0], VideoStreamInfo):
                raise RuntimeError(
                    "expected exactly one video stream from the encode session's fMP4"
                )

            muxer = Muxer()
            video_track = muxer.add_video_track(streams[0])
            audio_track = muxer.add_audio_track(audio_info)
            live = muxer.begin()
            while True:
                packet = demuxer.poll_packet()
                if packet is None:
                    break
                live.push_packet(dataclasses.replace(packet, stream_index=video_track))
            for packet in audio_packets:
                live.push_packet(dataclasses.replace(packet, stream_index=audio_track))
            live.flush()
            mp4 = live.poll_bytes()

    out_path = Path("out.mp4")
    out_path.write_bytes(mp4)
    if have_audio:
        print(
            f"recorded {frames_recorded} frames over {RECORD_SECONDS:.0f} s + "
            f"{len(audio_packets)} AAC packets from {pcm_bytes} PCM bytes -> "
            f"{out_path} ({len(mp4)} bytes, two tracks)"
        )
    else:
        print(
            f"recorded {frames_recorded} frames over {RECORD_SECONDS:.0f} s "
            f"(audio unavailable) -> {out_path} ({len(mp4)} bytes, video only)"
        )


if __name__ == "__main__":
    main()
